ArgumentResolver/EntityValueResolver.php

```
@@ -0,0 +1,293 @@
+            return false;
+        }
+
+        $om = $this->getManager($options['entity_manager'], $class);
+        if ($options['evict_cache'] && $om instanceof EntityManagerInterface) {
+            $cacheProvider = $om->getCache();
+            if ($cacheProvider && $cacheProvider->containsEntity($class, $id)) {
+                $cacheProvider->evictEntity($class, $id);
+            }
+        }
+
+        try {
+            return $om->getRepository($class)->find($id);
+        } catch (NoResultException | ConversionException $e) {
+            return null;
+        }
+    }
+
+    /**
+     * @param array<string, string> $options
+     *
+     * @return false|mixed|mixed[]
+     */
+    private function getIdentifier(Request $request, array $options, string $name)
+    {
+        if ($options['id'] !== null) {
+            if (is_array($options['id'])) {
+                $id = [];
+                foreach ($options['id'] as $field) {
+                    // Convert "%s_uuid" to "foobar_uuid"
+                    if (strstr($field, '%s') !== false) {
+                        $field = sprintf($field, $name);
+                    }
+
+                    $id[$field] = $request->attributes->get($field);
+                }
+
+                return $id;
+            }
+
+            $name = $options['id'];
+        }
+
+        if ($request->attributes->has($name)) {
+            return $request->attributes->get($name);
+        }
+
+        if ($request->attributes->has('id') && ! $options['id']) {
+            return $request->attributes->get('id');
+        }
+
+        return false;
+    }
+
+    /**
+     * @param array<string, string> $options
+     *
+     * @return false|object|null
+     */
+    private function findOneBy(string $class, Request $request, array $options)
+    {
+        if (! $options['mapping']) {
+            $keys               = $request->attributes->keys();
+            $options['mapping'] = $keys ? array_combine($keys, $keys) : [];
```

This default mapping that takes all request attributes to find the entity is something that I'd like to make configurable (through the config, not in each argument), as it is quite confusing. I had several bugs in my project where an optional argument was suddenly given a weird value instead of `null` just because the controller action had another argument with a name matching the name of an association (and that this default mapping would silently use it to build a criteria)

---

providing a semantic configuration that will change the default behavior is something that could/should be done in a dedicated PR IMHO.

Maybe related to https://github.com/doctrine/DoctrineBundle/pull/1413#discussion_r736867822

---

I don't want to disable the whole argument resolver. I want a way to disable the "use any other request param as a lookup" behavior. In my project, I never found a case where I wanted it, and I got several bugs caused by it for optional parameters of controllers.
Cases that actually want to do a lookup by another argument than the current argument will often need to provide a mapping explicitly anyway, as they expect to use a given field, not "any field" either (maybe they are just lucky that the names of their other controller arguments don't match fields that exist in their entity today, and they won't add new fields or new arguments in the future that would break that lucky invariant)

---

Making this configurable globally would means bundles won't be able to rely on one the the behaviors. My take on this would be to completely remove the feature as something that is to much magic for an argument resolver.

---

well, I'm also fine with removing the magic mapping entirely too (the `mapping` option should still exist for cases where it is configured explicitly, as it solves a use case)

---

> use any other request param 

other than what? `id` and the attribute with the same name than the argument will both use the method `find` . here we are talking about the fallback that use attributes that match none of the 2 cases above.

Disabling this feature globally does not make sens. If the request does not contain `id` nor an attribute with the same, then there will be nothing to match, entity could not be found, and user would have to provide a mapping on every attributes anyway.

Sorry, I disagree with your 

`/blogs/{slug}` or `/product/{uuid}` just work, and is often used. No need for custom mapping in such cases.

At the opposite, I never had issue with route containing several attributes. maybe I'm lucky, or because routes with multiple arguments are rare enough to provide mapping or expr and are not ambiguous

---

anyway, I added a configuration section and added an `auto_mapping` flag that allows users to disable this behavior. This will enforce people to either use `{id}` (or the same name as the argument) OR to explicitly provide a `mapping` option (or `id` or `expr`).